Services/DevicePropertiesRequest.cs

```diff
@@ -0,0 +1,96 @@
+    }
+
+    public class DevicePropertiesRequest : IDevicePropertiesRequest
+    {
+        private readonly Azure.Devices.Client.DeviceClient client;
+        private readonly ILogger log;
+        private string deviceId;
+        private ISmartDictionary deviceProperties;
+
+        public DevicePropertiesRequest(Azure.Devices.Client.DeviceClient client, ILogger logger)
+        {
+            this.client = client;
+            this.log = logger;
+            this.deviceId = string.Empty;
+        }
+
+        public async Task RegisterDevicePropertyUpdatesAsync(string deviceId, ISmartDictionary deviceProperties)
+        {
+            if (this.deviceId != string.Empty)
+            {
+                this.log.Error("Application error, each device must have a separate instance", () => { });
+                throw new Exception("Application error, each device must have a separate instance of " + this.GetType().FullName);
+            }
+
+            this.deviceId = deviceId;
+            this.deviceProperties = deviceProperties;
+
+            this.log.Debug("Setting up callback for desired properties updates.", () => new { this.deviceId });
+
+            // Set callback that IoT Hub calls whenever the client receives a state update (desired or reported).
+            // This has the side-effect of subscribing to the PATCH topic on the service.
+            // https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.client.deviceclient.setdesiredpropertyupdatecallbackasync
+            await this.client.SetDesiredPropertyUpdateCallbackAsync(OnPropertyUpdateRequestedCallback, null);
+
+            this.log.Debug("Callback for desired properties updates setup successfully", () => new { this.deviceId });
+        }
+
+        /// <summary>
+        /// When a desired property change is requested, update the internal the device state properties
+        /// which will be reported to the hub. If there is a new desired property that does not exist in
+        /// the reported properties, it will be added.
+        /// </summary>
+        public Task OnPropertyUpdateRequestedCallback(TwinCollection desiredProperties, object userContext)
+        {
+            this.log.Info("Desired property update requested", () => new { this.deviceId, desiredProperties });
+
+            if (desiredProperties != null && desiredProperties.Count > 0)
+            {
+                // This is where custom code for handling specific desired property changes could be added.
+                // For the purposes of the simulation service, we have chosen to write the desired properties
+                // directly to the reported properties. 
+                try
+                {
+                    foreach (KeyValuePair<string, object> item in desiredProperties)
+                    {
+                        // Only update if key doesn't exist or value has changed 
+                        if (!this.deviceProperties.Has(item.Key) ||
```

Comment:
What happens if a new desired property is added?  this looks like it only updates existing values & doesn't add new properties, correct?

Reply:
`!this.deviceProperties.Has(item.Key)` will evaluate as true and `.Set()` will be called which has the behavior of updating existing properties OR creating new properties if it doesn't exist. (see comment below)

Also see unit test `SmartDictionary_Should_HaveNewItem_When_NewDesiredPropertyAdded()` in `DevicePropertiesRequestTest.cs`